pdg/src/main.rs

```diff
@@ -24,62 +24,111 @@ mod query;
 mod util;
 
 use builder::{construct_pdg, read_event_log};
-use clap::Parser;
+use c2rust_analysis_rt::metadata::WithMetadata;
+use clap::{Parser, ValueEnum};
 use color_eyre::eyre;
-use std::path::{Path, PathBuf};
+use std::{
+    collections::HashSet,
+    fmt::{self, Display, Formatter},
+    path::{Path, PathBuf},
+};
 
-use crate::{builder::read_metadata};
+use crate::{builder::read_metadata, graph::Func};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, ValueEnum)]
+enum ToPrint {
+    Graphs,
+    Counts,
+    Events,
+    LatestAssignments,
+    WritePermissions,
+}
+
+impl Display for ToPrint {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(self.to_possible_value().unwrap().get_name())
+    }
+}
 
 /// Construct and query a PDG from an instrumented program's event log.
-#[derive(Parser, Debug)]
+#[derive(Debug, Parser)]
 #[clap(author, version, about, long_about = None)]
 struct Args {
     /// Path to an event log from a run of an instrumented program.
     #[clap(long, value_parser)]
     event_log: PathBuf,
+
     /// Path to the instrumented program's metadata generated at compile/instrumentation time.
     #[clap(long, value_parser)]
     metadata: PathBuf,
+
+    /// What to print.
+    #[clap(long, value_parser)]
+    print: Vec<ToPrint>,
 }
 
 fn main() -> eyre::Result<()> {
     color_eyre::install()?;
     env_logger::init();
     let args = Args::parse();
 
+    let print = args.print.iter().collect::<HashSet<_>>();
+    let print = |to_print| print.contains(&to_print);
```

Comment:
I'm curious how the shadowing and closing over `print` works here, does it create a local copy in the lambda?

Reply:
it references and shadows the existing `print` (it would move if it were a `move` closure). That said, I think `print_args` and `should_print` would be more descriptive names.

Reply:
Okay, I'll change the variable names; that's clearer.

Reply:
Fixed in https://github.com/immunant/c2rust/pull/493/commits/1d5baf6cbe1a7730bcfdbb03434d02e097ccfcce.